components/script/dom/idbtransaction.rs

```diff
@@ -0,0 +1,322 @@
+        // Step 1: Handle the case where transaction has finised
+        if self.finished.get() {
+            return Err(Error::InvalidState);
+        }
+
+        // Step 2: Check that the object store exists
+        if !self.object_store_names.Contains(name.clone()) {
+            return Err(Error::NotFound);
+        }
+
+        // Step 3: Each call to this method on the same
+        // IDBTransaction instance with the same name
+        // returns the same IDBObjectStore instance.
+        let mut store_handles = self.store_handles.borrow_mut();
+        let store = store_handles.entry(name.to_string()).or_insert({
+            let store =
+                IDBObjectStore::new(&self.global().as_window(), self.db.get_name(), name, None);
+            store.set_transaction(&self);
+            Dom::from_ref(&*store)
+        });
+
+        Ok(DomRoot::from_ref(&*store))
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#commit-transaction
+    fn Commit(&self) -> Fallible<()> {
+        // Step 1
+        let (sender, receiver) = ipc::channel(self.global().time_profiler_chan().clone()).unwrap();
+        let start_operation = SyncOperation::Commit(
+            sender,
+            self.global().origin().immutable().clone(),
+            self.db.get_name().to_string(),
+            self.serial_number,
+        );
+
+        self.get_idb_thread()
+            .send(IndexedDBThreadMsg::Sync(start_operation))
+            .unwrap();
+
+        let result = receiver.recv().unwrap();
+
+        // Step 2
+        if let IndexedDBThreadReturnType::Commit(Err(_result)) = result {
+            // FIXME:(rasviitanen) also support Unknown error
+            return Err(Error::QuotaExceeded);
+        }
+
+        // Step 3
+        // FIXME:(rasviitanen) https://www.w3.org/TR/IndexedDB-2/#commit-a-transaction
+
+        // Steps 3.1 and 3.3
+        self.dispatch_complete();
+
+        Ok(())
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbtransaction-abort
+    fn Abort(&self) -> Fallible<()> {
+        // FIXME:(rasviitanen)
+        // This only sets the flags, and does not abort the transaction
+        // see https://www.w3.org/TR/IndexedDB-2/#abort-a-transaction
+        if self.finished.get() {
+            return Err(Error::InvalidState);
+        }
+
+        self.active.set(false);
+
+        Ok(())
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbtransaction-objectstorenames
+    fn ObjectStoreNames(&self) -> DomRoot<DOMStringList> {
+        self.object_store_names.clone()
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbtransaction-mode
+    fn Mode(&self) -> IDBTransactionMode {
+        self.mode
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbtransaction-error
+    fn Error(&self) -> DomRoot<DOMException> {
+        // FIXME:(rasviitanen) ???
+        // It's weird that the WebIDL specifies that this isn't returning an Option.
```

Comment:
Filed https://github.com/w3c/IndexedDB/issues/312. Gecko uses ?, so we should as well.